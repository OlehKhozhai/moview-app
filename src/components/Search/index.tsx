import React from 'react';
import cn from 'classnames';
import { useHistory } from 'react-router-dom';

import Button from 'components/_common/Button';
import { setSearchParams } from 'helpers/common';
import styles from './styles.module.scss';

type SearchProps = {
  className?: string;
};

const Search: React.FC<SearchProps> = ({ className }) => {
  const [searchValue, setSearchValue] = React.useState('');
  const history = useHistory();

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
  };

  const handleSearchButtonClick = React.useCallback(() => {
    if (searchValue) {
      const searchParams = setSearchParams({
        query: history.location.search,
        params: [
          { key: 'search', value: searchValue },
          { key: 'searchBy', value: 'title' },
        ],
      });
      history.replace({ search: searchParams });

      setSearchValue('');
    }
  }, [searchValue, history]);

  React.useEffect(() => {
    const handleKeyPress = (e: KeyboardEvent) => {
      if (e.code === 'Enter') {
        handleSearchButtonClick();
      }
    };

    document.addEventListener('keypress', handleKeyPress);

    return () => {
      document.removeEventListener('keypress', handleKeyPress);
    };
  }, [handleSearchButtonClick]);

  return (
    <div className={cn(styles.root, className)}>
      <input
        placeholder="What do you want to watch?"
        className={styles.input}
        value={searchValue}
        onChange={handleInputChange}
      />
      <Button
        title="Search"
        className={styles.button}
        disabled={!searchValue}
        onClick={handleSearchButtonClick}
      />
    </div>
  );
};

export default React.memo(Search);
