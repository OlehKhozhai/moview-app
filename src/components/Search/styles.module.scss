@import 'assets/styles/mixins/colors';

.root {
  display: flex;
  align-items: center;
  justify-content: space-between;
  z-index: 1;
  position: inherit;
}

.input{
  width: 100%;
  background-color: $input-primary-light-bg;
  margin-right: 10px;
  padding: 18px 20px;
  border-radius: 5px;
  color: $primary-color;
}

.button {
  padding: 15px 75px;
}

