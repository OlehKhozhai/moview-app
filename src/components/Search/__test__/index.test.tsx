import React from 'react';
import { create } from 'react-test-renderer';

import Search from '../index';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn().mockReturnValue({
    replace: jest.fn(),
    location: {
      search: '',
    },
  }),
}));

describe('Search component', () => {
  test('should render markup by default', () => {
    const component = create(<Search />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should handle input changes', () => {
    const setSearchValueStub = jest.fn();
    jest.spyOn(React, 'useState').mockReturnValue(['', setSearchValueStub]);
    const component = create(<Search />);

    component.root.findByType('input').props.onChange({ target: { value: 'test' } });

    expect(setSearchValueStub).toHaveBeenCalledWith('test');
  });

  test('should clear input value after click on button', () => {
    const setSearchValueStub = jest.fn();
    jest.spyOn(React, 'useState').mockReturnValue(['test', setSearchValueStub]);
    const component = create(<Search />);

    component.root.findByType('button').props.onClick();

    expect(setSearchValueStub).toHaveBeenCalledWith('');
  });
});
