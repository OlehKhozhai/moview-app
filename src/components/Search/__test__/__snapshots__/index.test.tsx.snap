// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render markup by default 1`] = `
<div
  className="root"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="What do you want to watch?"
    value=""
  />
  <button
    className="root primary button"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Search
  </button>
</div>
`;
